File: python/Remove Nth Node From End of List.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head, n):
        def index(node):
            if not node:
                return 0
            i = index(node.next) + 1
            if i > n:
                node.next.val = node.val
            return i

        index(head)
        return head.next

# Example usage:
# Create your linked list
head = ListNode(1)
head.next = ListNode(2)

# Test the removeNthFromEnd method
solution = Solution()
result = solution.removeNthFromEnd(head, 1)

# Print the result or perform any other necessary operations
while result:
    print(result.val, end=" ")
    result = result.next

File: JAVA/java/src/RemoveNthNodeFromEndofList.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
    }
}

public class RemoveNthNodeFromEndofList {
    public ListNode removeNthFromEnd(ListNode head, int n) {

        ListNode start = new ListNode(0);
        ListNode slow = start, fast = start;
        start.next = head;

        // Move fast pointer to the (n + 1)-th node from the beginning
        for (int i = 0; i <= n; i++) {
            fast = fast.next;
        }

        // Move both pointers until fast reaches the end
        while (fast != null) {
            slow = slow.next;
            fast = fast.next;
        }

        // Remove the N-th node from the end
        slow.next = slow.next.next;

        return start.next;
    }

    public static void main(String[] args) {
        // Example usage:
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);

        RemoveNthNodeFromEndofList ob = new RemoveNthNodeFromEndofList();
        ListNode result = ob.removeNthFromEnd(head, 1);

        // Print the result or perform any other necessary operations
        while (result != null) {
            System.out.print(result.val + " ");
            result = result.next;
        }
    }
}

